KFSPTS-9246: Changes for creation of Domestic DV Vendors from PaymentWorks to KFS PVEN.

nkimble opened the pull request:
This PR relies upon the SQL changes in PR #249 and must be merged at the same time that one is merged.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/dataaccess/impl/PaymentWorksIsoFipsCountryDaoJdbc.java:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+    
+    private String buildIsoToFipsCountryTranslationsSql() {
```
Do we need all these mappings in memory? This code will be more difficult to maintain than code that is using the object/relational mapping.

nkimble replied:
The data that is received by KFS via the PaymentWorks integration is not validated at all.  So, in the case of countries, validation and conversion from ISO to FIPS needs to be performed for 5 countries for every DV vendor and 6 countries for every PO vendor. I am not sure about RV vendors yet. With that being said, I am concerned about the PaymentWorks batch jobs becoming database bound for all of that validation/data conversion. That is the reason why I chose the read the database once into a Map design.

db483 replied:
Thank you for explaining this. I am wondering if in this case it might make sense to use a view so that we can have relational mapping defined in ojb and make the code more clean an easy to maintain.

If you decide against a view I think it would be good to have the SQL defined in a constant.

nkimble replied:
Per our offline conversation, leaving the Map implementation and moving the SQL to a constant at the top of the class.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksDtoToPaymentWorksVendorConversionServiceImpl.java:
```diff
@@ -0,0 +1,246 @@
+    }
+    
+    private void extractCustomFields(PaymentWorksVendor stgNewVendor, PaymentWorksNewVendorRequestDetailDTO pmwNewVendorDetailDTO, PaymentWorksNewVendorRequestsBatchReportData reportData) {
+        Map<String, String> customFieldsDTOMap = createCustomFieldIdCustomFieldValueMap(pmwNewVendorDetailDTO.getCustom_fields());
```
this method is quite complicated, it could probably be broken down into smaller units

nkimble replied:
Broke method up as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsReportServiceImpl.java:
```diff
@@ -0,0 +1,486 @@
+        LOG.info("sendResultsEmail: entered");
+        String body = readReportFileToString(reportData, reportFile);
+        String subject = buildEmailSubject(reportData);
+        if (StringUtils.isNotBlank(getToAddress()) && StringUtils.isNotBlank(getFromAddress()) && StringUtils.isNotBlank(body) && StringUtils.isNotBlank(subject)) {
```
It would probably be better to do all these checks and logging inside the sendEmail method.

nkimble replied:
Moved as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsServiceImpl.java:
```diff
@@ -0,0 +1,412 @@
+                        getPaymentWorksWebServiceCallsService().sendApprovedStatusToPaymentWorksForNewVendor(savedStgNewVendorRequestDetailToProcess.getPmwVendorRequestId());
+                    }
+                    else {
+                        getPaymentWorksWebServiceCallsService().sendRejectedStatusToPaymentWorksForNewVendor(savedStgNewVendorRequestDetailToProcess.getPmwVendorRequestId());
```
It might make sense to have a newVendorSuccessful boolean to keep track of whether the vendor was accepted or rejected. Then you only need to call the sendRejectedStatusToPaymentWorksForNewVendor method once and the code for this method will be much shorter.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsServiceImpl.java:
```diff
@@ -0,0 +1,412 @@
+    }
+    
+    private boolean pmwDtosCouldConvertCustomAttributesToPmwJavaClassAttributes(PaymentWorksVendor stgNewVendorRequestDetailToProcess) {
+        if (stgNewVendorRequestDetailToProcess.isCustomFieldConversionErrors()) {
```
could probably just return !stgNewVendorRequestDetailToProcess.isCustomFieldConversionErrors()?

nkimble replied:
Changed as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsServiceImpl.java:
```diff
@@ -0,0 +1,412 @@
+    
+    private boolean pmwNewVendorAttributesConformToKfsLengthsOrFormats(PaymentWorksVendor stgNewVendorRequestDetailToProcess, List<String> errorMessages) {
+        boolean allValidationPassed = true;
+        if (StringUtils.isNotBlank(stgNewVendorRequestDetailToProcess.getRequestingCompanyLegalLastName()) &
```
could this be extracted in a separate method?

nkimble replied:
Changed as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsServiceImpl.java:
```diff
@@ -0,0 +1,412 @@
+            }
+        }
+        else {
+            if (StringUtils.isNotBlank(stgNewVendorRequestDetailToProcess.getRequestingCompanyLegalName())) {
```
could this be extracted in a separate method?

nkimble replied:
Changed as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsServiceImpl.java:
```diff
@@ -0,0 +1,412 @@
+    
+    private PaymentWorksVendor pmwPendingNewVendorRequestInitialSaveToKfsStagingTable(PaymentWorksVendor pmwVendor) {
+        LOG.info("pmwPendingNewVendorRequestSavesToKfsStagingTableSuccessful: entered");
+        pmwVendor.setPmwRequestStatus(PaymentWorksConstants.PaymentWorksNewVendorRequestStatusType.PENDING.getText());
```
the part where the pmwVendor fields are updated could probably be extracted in a separate method

nkimble replied:
Setting status values moved into new method as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsServiceImpl.java:
```diff
@@ -0,0 +1,412 @@
+    
+    private void updatePmwNewVendorStagingTableBasedOnSuccessfulCreateVendorProcessing(PaymentWorksVendor pmwVendor) {
+        LOG.info("updatePmwNewVendorStagingTableBasedOnSuccessfulCreateVendorProcessing: entered");
+        pmwVendor.setPmwRequestStatus(PaymentWorksConstants.PaymentWorksNewVendorRequestStatusType.APPROVED.getText());
```
same here

nkimble replied:
Setting status values moved into new method as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsServiceImpl.java:
```diff
@@ -0,0 +1,412 @@
+    
+    private void updatePmwNewVendorStagingTableBasedOnCreateVendorFailing(PaymentWorksVendor pmwVendor) {
+        LOG.info("updatePmwNewVendorStagingTableBasedOnCreateVendorFailing: entered");
+        pmwVendor.setPmwRequestStatus(PaymentWorksConstants.PaymentWorksNewVendorRequestStatusType.REJECTED.getText());
```
same here

nkimble replied:
Setting status values moved into new method as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksVendorToKfsVendorDetailConversionServiceImpl.java:
```diff
@@ -0,0 +1,621 @@
+    protected KfsVendorDataWrapper createVendorDetail(PaymentWorksVendor pmwVendor, 
+                                                      Map<String, List<PaymentWorksIsoFipsCountryItem>> paymentWorksIsoToFipsCountryMap, 
+                                                      Map<String, SupplierDiversity> paymentWorksToKfsDiversityMap) {
+        //REMOVE THIS IF-STATEMENT WHEN FIRST PO IMPLEMENTATION IS DONE. 
```
perhaps this should be removed altogether, or a method that does nothing could be added for the PO implementation. It would probably be good to have a separate explicit method for the Domestic DV vendors implementation?

nkimble replied:
Per our offline discussion, PO vendors contain all of the data that comprise DV vendors plus more data. This if-statement is present to short circuit processing right away if the vendor is detected to be a PO vendor because PaymentWorks could send a PO vendor's data to this initial batch job implementation that is not equipped to handle it. I have added a more descriptively named method and removed the comments.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksVendorToKfsVendorDetailConversionServiceImpl.java:
```diff
@@ -0,0 +1,621 @@
+        switch (determineTaxRuleToUseForDataPopulation(pmwVendor, convertIsoCountryCodeToFipsCountryCode(pmwVendor.getRequestingCompanyTaxCountry(), paymentWorksIsoToFipsCountryMap))) {
+            case PaymentWorksConstants.PaymentWorksNewVendorTaxBusinessRule.INDIVIDUAL_US_SSN: 
+            {
+                vendorHeader.setVendorForeignIndicator(new Boolean(false));
```
the code for each of these cases should be extracted in separate methods

nkimble replied:
Separate methods created as requested.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksVendorToKfsVendorDetailConversionServiceImpl.java:
```diff
@@ -0,0 +1,621 @@
+        allVendorAddresses.add(buildTaxAddress(pmwVendor, paymentWorksIsoToFipsCountryMap));
+        allVendorAddresses.add(buildRemitAddress(pmwVendor, paymentWorksIsoToFipsCountryMap));
+        
+        // WHEN FIRST PO IMPLEMENTATION IS DONE ADD THIS CONDITIONAL BACK INTO CODE.
```
we shouldn't have these kind of comments in the code. Perhaps we could find a better way to deal with this, like a method for PO that now does nothing?

nkimble replied:
Removed comments

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksVendorToKfsVendorDetailConversionServiceImpl.java:
```diff
@@ -0,0 +1,621 @@
+                pmwVendor.getVendorInformationContactName(), pmwVendor.getVendorInformationEmail(), 
+                pmwVendor.getVendorInformationPhoneNumber(), pmwVendor.getVendorInformationPhoneExtension()));
+        
+        //ADD THESE CONTACT TYPES WHEN PO VENDOR IMPLEMENTATION IS ADDED
```
again, these comments should be removed

nkimble replied:
Removed.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksVendorToKfsVendorDetailConversionServiceImpl.java:
```diff
@@ -0,0 +1,621 @@
+    }
+
+    private int determineTaxRuleToUseForDataPopulation(PaymentWorksVendor pmwVendor, String pmwVendorFipsTaxCountryCode) {
+        if (isPmwVendorIndividualSolePropriatorSingleMemberLlc(pmwVendor) && 
```
perhaps the checks could be extracted into a method called isIndividualUsSsn? or something along those lines? same goes for the other checks in that determine the tax rule.

nkimble replied:
Extracted each check into its own method.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksVendorToKfsVendorDetailConversionServiceImpl.java:
```diff
@@ -0,0 +1,621 @@
+    }
+
+    //CREATE SIMILAR ROUTINE FOR W8 WHEN FORIEGN VENDOR PROCESSING IS ADDED
+    private KfsVendorDataWrapper populateW9Attributes(KfsVendorDataWrapper kfsVendorDataWrapper, PaymentWorksVendor pmwVendor) {
```
should this method be called populateW9AttributesForDomesticVendor? and instead of the comment maybe have a method pupulateW9AttributesForForeignVendor? Although it is probably best to have the current code only deal with the current implementation.

nkimble replied:
Removed the comment.

db483 commented on src/main/resources/edu/cornell/kfs/pmw/batch/businessobject/datadictionary/PaymentWorksVendor.xml:
```diff
@@ -0,0 +1,1511 @@
+        <property name="forceUppercase" value="false" />
+        <property name="maxLength" value="200" />
+        <property name="control">
+            <bean parent="TextControlDefinition" p:size="50" />
```
should the size be 200 in this case? or at least a bit longer?

db483 commented on src/main/resources/edu/cornell/kfs/pmw/cu-ojb-pmw.xml:
```diff
@@ -0,0 +1,172 @@
+    <field-descriptor name="kfsAchProcessingStatus" column="KFS_ACH_PROC_STAT" jdbc-type="VARCHAR" />\
+    <field-descriptor name="processTimestamp" column="PROC_TS" jdbc-type="TIMESTAMP"/>
+
+    <field-descriptor name="requestingCompanyId" column="REQ_CMP_ID" jdbc-type="VARCHAR" />
```
I wonder if it would have made sense for this to be a separate BO

nkimble replied:
Per our offline conversation, this will not be broken into another BO as this data is all part of the single PMW form submitted by a single vendor.  Breaking it into another business object would make tracking the data submitted more difficult.

db483 commented:
There is a lot of code in the current PR and quite a good chunk of it is not related to the current implementation. I think it would be good to remove that code from this PR. It will make it easier for you to manage the code and requested changes - I would only expect the concerns for the code related to Domestic Vendors to be addressed with this PR. It would reduce both the time you spend getting the code ready to be merged and the time it takes the reviewer to look at the code. I believe it will help us both focus on the current implementation. Thank you!

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+    public final static String SIMPLE_DATE_FORMAT_REPRESENTATION_FOR_DATE = "MM/dd/yyyy";
+    public final static String DIVERSITY_EXPIRATION_DATE_CERTIFIED = "Certified";
+    public final static String REGEX_FOR_MM_SLASH_DD_SLASH_YYYY = "^(1[0-2]|0[1-9])/(3[01]|[12][0-9]|0[1-9])/[0-9]{4}$";
+    public final static Pattern PATTERN_COMPILED_REGEX_FOR_MM_SLASH_DD_SLASH_YYYY = Pattern.compile(REGEX_FOR_MM_SLASH_DD_SLASH_YYYY);
```
I think some of the files were not loaded in the PR and that is why I could not see where this was used. I can see now that it's used.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+        public static final String PMW_VENDOR_REQUEST_ID = "pmwVendorRequestId";
+    }
+    
+    public static final class PaymentWorksFieldMappingDatabaseFieldNames {
```
I am not sure I understand why the ojb mapping isn't used...

nkimble replied:
Adjusted the code for this.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+        public static final String VENDOR_INFORMATION_FORM = "VI";
+    }
+    
+    public static final class KFSVendorContactPhoneTypes {
```
Found these too..

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+    }
+    //Constants were declared instead of using the values directly in the 
+    //enumeration to allow the use of switch statements in the code.
+    public static final int INDIVIDUAL_SOLE_PROPRIETOR_TAX_CLASSIFICATION_INDICATOR  = 0;
```
found these too, so ignore the comment

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+
+public class PaymentWorksConstants {
+    
+    public static final String OUTPUT_ATTRIBUTE_BEGIN_DELIMITER = ":<";
```
It appears refactoring of the Concur code has removed these delimiters. What delimiter would you suggest that already exists in the code base to wrap a value in a log file.  The intent of these delimiters is to show there exists leading/trailing white space or that the value is empty/blank. Things you would not readily see in a log file if you just output the value.

db483 replied:
I see, I was remembering these delimiters from a previous PR but did not check that they are still used in Concur. Interesting that they were removed with the refactoring of the Concur code.

nkimble replied:
Would you prefer some other existing delimiter used instead of these?

db483 replied:
I don't think a good delimiter exists for the purpose. I remembered seeing this in the Concur code and thought it was a good idea to have a common constant that perhaps other future code could use.

db483 commented:
It looks like several files were not initially loaded in git as I was reviewing this PR so I have checked out the branch locally to review this again. I think they were not loaded because they had a large amount of changes.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/dataaccess/impl/KfsSupplierDiversityDaoJdbc.java:
```diff
@@ -0,0 +1,81 @@
+    private static final String PMW_SUPP_DVRST_DESC_COL = "pmw_supp_dvrst_desc";
+
+    @Override
+    public Map<String, SupplierDiversity> buildPmwToKfsSupplierDiversityMap() {
```
I found this so ignore the comment about not being used.

nkimble replied:
Per our offline conversation, leaving the Map implementation and moving the SQL to a constant at the top of the class.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/dataaccess/impl/PaymentWorksIsoFipsCountryDaoJdbc.java:
```diff
@@ -0,0 +1,86 @@
+    private static final String FIPS_CNTRY_NM_COL = "fips_cntry_nm";
+
+    @Override
+    public Map<String, List<PaymentWorksIsoFipsCountryItem>> buildIsoToFipsMapFromDatabase() {
```
I found this so ignore the comment about not being used.

nkimble replied:
Per our offline conversation, leaving the Map implementation and moving the SQL to a constant at the top of the class.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/service/PaymentWorksDataProcessingIntoKfsService.java:
```diff
@@ -0,0 +1,17 @@
+
+public interface PaymentWorksDataProcessingIntoKfsService {
+    
+    boolean ableToCreateValidateRouteKfsVendor(PaymentWorksVendor savedStgNewVendorRequestDetailToProcess, Map<String, List<PaymentWorksIsoFipsCountryItem>> paymentWorksIsoToFipsCountryMap,
```
ignore this one

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+    public static final String OUTPUT_ATTRIBUTE_END_DELIMITER = "> ";
+    public static final String OUTPUT_RESTRICTED_DATA_PRESENT = "RestrictedDataPresent";
+
+    public final static String SIMPLE_DATE_FORMAT_REPRESENTATION_FOR_DATE = "MM/dd/yyyy";
```
Yes, format is the same. Good catch.

db483 commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+        public static final String PAYMENTWORKS_NEW_VENDOR_CREATE_ROUTE_ANNOTATION = "Generated by PaymentWorks New Vendor Create Batch Process";
+    }
+    
+    public static final class KFSVendorContactTypes {
```
ignore the comment about not being referenced

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+    }
+    
+    public static final class PaymentWorksStagingTableColumnConstants {
+        public static final String PMW_VENDOR_REQUEST_ID = "pmwVendorRequestId";
```
Moving as suggested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+    }
+    
+    public static final class KFSVendorMaintenaceDocumentConstants {
+        public static final String DESCRIPTION_SUFFIX = "--new vendor";
```
Changing as suggested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+    }
+    
+    public static final class KFSNotesStrings {
+        public static final class TEXT_FOR_PVEN_NOTES {
```
Moved as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+        }
+    }
+    
+    public static final class PaymentWorksBatchReportMessages {
```
Moved as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksConstants.java:
```diff
@@ -0,0 +1,245 @@
+        public static final String NEW_VENDOR_REQUESTS_REPORT_NAME = "New Vendor Requests Create KFS Vendor";
+    }
+    
+    public static final class PaymentWorksBusinessRuleFailureMessages {
```
Moved as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksParameterConstants.java:
```diff
@@ -0,0 +1,28 @@
+package edu.cornell.kfs.pmw.batch;
+
+public class PaymentWorksParameterConstants {
```
I am confused by this request. My intent was to model existing code. Both base code and our customized code have separate files where parameter constants are defined. Why should the PaymentWorks mod be different?

db483 replied:
Looking at base code this is not consistent at all, there are parameters in KFSConstants file, there are parameters in KFSParametersKeyConstants files and occasionally there are parameters in parameter constants files for specific modules. We seem to more consistently place parameter constants in a separate file in cu-kfs so I guess your point makes sense.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/PaymentWorksParameterConstants.java:
```diff
@@ -0,0 +1,28 @@
+package edu.cornell.kfs.pmw.batch;
+
+public class PaymentWorksParameterConstants {
+    public static final String PAYMENTWORKS_PARAMETER_NAMESPACE = "KFS-PMW";
```
Moving namespace as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/dataaccess/impl/PaymentWorksVendorDaoJdbc.java:
```diff
@@ -0,0 +1,92 @@
+    
+    @Override
+    public boolean isExistingPaymentWorksVendor(String pmwVendorId) {
+        Map<String, String> fieldValues = new HashMap<String, String>();
```
How to determine whether the vendor received from PaymentWorks is a duplicate request is to see whether that vendor's PMW ID already exists in the database where the PMW vendors previously processed are held.  The work performed by this method to check the existence of that vendor in the database would still need to to be coded... under a different name.  Why would we want to add more overhead by having an if-check in a service class that would then call the code in this method?

db483 replied:
I am not sure I understand this. Is this method still to be changed to something that would call the database directly through jdbc?

Having a service defined follows the guidelines for KFS coding standards and application design patterns:
https://kualico.atlassian.net/wiki/spaces/FINDOC/pages/1671258/Java+Coding+Standards#JavaCodingStandards-ApplicationArchitectureandDesignPatterns

I might be misunderstanding what you are trying to accomplish with this code. Please let me know if you would rather chat about this. Thank you!

nkimble replied:
Per our discussion, moved method isExistingPaymentWorksVendor to service class PaymentWorksNewVendorRequestsServiceImpl.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksDataProcessingIntoKfsServiceImpl.java:
```diff
@@ -0,0 +1,213 @@
+    protected ConfigurationService configurationService;
+    
+    @Override
+    public boolean ableToCreateValidateRouteKfsVendor(PaymentWorksVendor pmwVendor, Map<String, List<PaymentWorksIsoFipsCountryItem>> paymentWorksIsoToFipsCountryMap,
```
Renaming method as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksDataTransformationServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+        String returnValue = KFSConstants.EMPTY_STRING;
+        List<PaymentWorksTinType> matchingValues = Arrays.asList(PaymentWorksConstants.PaymentWorksTinType.values())
+                                                         .stream()
+                                                         .filter(tinType ->  tinType.getPmwCodeAsString().equalsIgnoreCase(pmwTinTypeCodeToConvert))
```
Fixed.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/service/impl/PaymentWorksNewVendorRequestsReportServiceImpl.java:
```diff
@@ -0,0 +1,486 @@
+        LOG.info("sendEmailThatNoDataWasFoundToProcess: entered");
+        String body = getPaymentWorksBatchUtilityService().retrievePaymentWorksParameterValue(PaymentWorksParameterConstants.PAYMENTWORKS_NEW_VENDOR_REPORT_NO_PENDING_VENDORS_FOUND_EMAIL_BODY);
+        String subject = getPaymentWorksBatchUtilityService().retrievePaymentWorksParameterValue(PaymentWorksParameterConstants.PAYMENTWORKS_NEW_VENDOR_REPORT_NO_PENDING_VENDORS_FOUND_EMAIL_SUBJECT);
+        if (StringUtils.isNotBlank(toAddress) && StringUtils.isNotBlank(fromAddress) && StringUtils.isNotBlank(body) && StringUtils.isNotBlank(subject)) {
```
Refactored same as above.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/dataaccess/impl/PaymentWorksVendorDaoJdbc.java:
```diff
@@ -0,0 +1,92 @@
+    
+    @Override
+    public PaymentWorksVendor savePaymentWorksVendorToStagingTable(PaymentWorksVendor pmwVendorToSave) {
+        pmwVendorToSave.setProcessTimestamp(getDateTimeService().getCurrentTimestamp());
```
Per our discussion, moved method savePaymentWorksVendorToStagingTable to service class PaymentWorksNewVendorRequestsServiceImpl.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/xmlObjects/PaymentWorksBankAccountDTO.java:
```diff
@@ -0,0 +1,140 @@
+public class PaymentWorksBankAccountDTO {
+
+    private String id;
+    private String bank_name;
```
Changed as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/xmlObjects/PaymentWorksCustomFieldDTO.java:
```diff
@@ -0,0 +1,57 @@
+@XmlAccessorType(XmlAccessType.FIELD)
+public class PaymentWorksCustomFieldDTO {
+
+    private String field_id;
```
Changed as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/xmlObjects/PaymentWorksCustomFieldsDTO.java:
```diff
@@ -0,0 +1,25 @@
+@XmlAccessorType(XmlAccessType.FIELD)
+public class PaymentWorksCustomFieldsDTO {
+
+    @XmlElement(name = "list-item")
```
Changed as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/xmlObjects/PaymentWorksNewVendorRequestDTO.java:
```diff
@@ -0,0 +1,30 @@
+public class PaymentWorksNewVendorRequestDTO {
+    
+    private String id;
+    private String request_status;
```
Changed as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/xmlObjects/PaymentWorksNewVendorRequestDetailDTO.java:
```diff
@@ -0,0 +1,53 @@
+    private String request_status;
+    @XmlElement(name = "custom_fields")
+    private PaymentWorksCustomFieldsDTO custom_fields;
+    private PaymentWorksRequestingCompanyDTO requesting_company;
```
Changed as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/xmlObjects/PaymentWorksRemittanceAddressDTO.java:
```diff
@@ -0,0 +1,23 @@
+@XmlAccessorType(XmlAccessType.FIELD)
+public class PaymentWorksRemittanceAddressDTO extends PaymentWorksAddressBaseDTO {
+
+    private PaymentWorksBankAccountDTO bank_acct;
```
Changed as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/xmlObjects/PaymentWorksRemittanceAddressesDTO.java:
```diff
@@ -0,0 +1,27 @@
+public class PaymentWorksRemittanceAddressesDTO {
+
+    @XmlElement(name = "list-item")
+    private List<PaymentWorksRemittanceAddressDTO> remittance_address;
```
Changed as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/xmlObjects/PaymentWorksRequestingCompanyDTO.java:
```diff
@@ -0,0 +1,178 @@
+public class PaymentWorksRequestingCompanyDTO {
+
+    private String id;
+    private String legal_name;
```
Changed as requested.

nkimble commented on src/main/java/edu/cornell/kfs/pmw/batch/dataaccess/impl/PaymentWorksVendorDaoJdbc.java:
```diff
@@ -0,0 +1,92 @@
+    }
+    
+    private String buildUpdateExistingPaymentWorksVendorInStagingTableSql(String pmwVendorRequestId, String pmwRequestStatus, String kfsVendorProcessingStatus, String kfsAchProcessingStatus, String kfsVendorDocumentNumber, Timestamp  processingTimeStamp) {
+        StringBuilder sql = new StringBuilder();
```
Per our offline conversation, this update SQL is to just maintain the status values pertaining to the KFS edoc processing of the PMW vendor data retrieved by the webservice. The status values for the vendor should be the ONLY data that is modified. Using the businessObjectService runs the risk of also modifying the PMW data.

nkimble commented:
Per our offline discussion today, I have:
* reverted the DTO changes to get the batch job to marshal the PMW returned web service data correctly
* reverted the PaymentWorksVendorDaoJdbc changes because of the increased maintenance that would have resulted.

db483 commented:
retest this please

nkimble commented:
retest this please

db483 commented:
retest this please

db483 commented:
retest this please

db483 commented:
retest this please

nkimble commented:
retest this please
